### PR TITLE
Refactor and fixes

### PR DESCRIPTION
**Pull request by Eraden**
Fixes:

* Fix wrong resolve merge commit
* Fix change log level

Refactor:
* Remove card controller

Improvements:
* Skip write modulation if temperature didn't changed
* Add more logs
* Add verbose monitor

### CODE REVIEW

**Comment by BoostCookie** [comment]
I will take a look at it on Sunday evening when I'm back at my computer.

**Comment by r15ch13** [comment]
Works great so far 👍🏽

File: src/main.rs [hunk 1]
```diff
@@ -19,6 +22,24 @@ pub enum AmdFanError {
...
     FailedReadVendor,
+    InvalidMonitorFormat,
+}
+
+impl std::fmt::Display for AmdFanError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            AmdFanError::InvalidPrefix => f.write_str("Card must starts with `card`."),
+            AmdFanError::InputTooShort => f.write_str(
+                "Card must starts with `card` and ends with number. Given name is too short.",
```
Comment (r15ch13) [inline src/main.rs:33]:
```suggestion
                "Card must start with `card` and ends with a number. The given name is too short.",
```

File: src/main.rs [hunk 2]
```diff
@@ -271,47 +348,36 @@ fn read_cards() -> std::io::Result<Vec<Card>> {
...
+        for hw_mon in controllers.iter_mut() {
+            let gpu_temp = hw_mon.max_gpu_temp().unwrap_or_default();
+            log::info!("Current {} temperature: {}", hw_mon.card, gpu_temp);
+            let last = *cache.entry(*hw_mon.card).or_insert(1_000f64);
 
-            let speed = config.speed_for_temp(gpu_temp);
-            let target_pwm = (speed as f32 * 2.55).round() as u32;
-            if controller.hw_mon.pwm_min() > target_pwm || controller.hw_mon.pwm_max() < target_pwm {
+            if ((last - 0.001f64)..(last + 0.001f64)).contains(&gpu_temp) {
+                log::info!("temperature didn't changed");
```
Comment (r15ch13) [inline src/main.rs:371]:
```suggestion
                log::info!("Temperature didn't change");
```

Would consider changing this to `log::debug` so it doesn't spam the system log.

File: src/main.rs [hunk 3]
```diff
@@ -271,47 +348,36 @@ fn read_cards() -> std::io::Result<Vec<Card>> {
...
 
 fn service(config: Config) -> std::io::Result<()> {
     let mut controllers = controllers(&config, true)?;
+    let mut cache = std::collections::HashMap::new();
     loop {
-        for controller in controllers.iter_mut() {
-            let gpu_temp = controller.hw_mon.gpu_temp().unwrap_or_default();
+        for hw_mon in controllers.iter_mut() {
+            let gpu_temp = hw_mon.max_gpu_temp().unwrap_or_default();
+            log::info!("Current {} temperature: {}", hw_mon.card, gpu_temp);
```
Comment (r15ch13) [inline src/main.rs:367]: Would consider changing this to `log::debug` so it doesn't spam the system log.

File: src/main.rs [hunk 4]
```diff
@@ -271,47 +348,36 @@ fn read_cards() -> std::io::Result<Vec<Card>> {
...
-            let speed = config.speed_for_temp(gpu_temp);
-            let target_pwm = (speed as f32 * 2.55).round() as u32;
-            if controller.hw_mon.pwm_min() > target_pwm || controller.hw_mon.pwm_max() < target_pwm {
+            if ((last - 0.001f64)..(last + 0.001f64)).contains(&gpu_temp) {
+                log::info!("temperature didn't changed");
                 continue;
-            }
+            };
+            let speed = config.speed_for_temp(gpu_temp);
+            log::info!("Resolved speed {:.2}", speed);
```
Comment (r15ch13) [inline src/main.rs:375]: Or combine all three log entries into one.

**Review by r15ch13 — APPROVED** [review]

File: src/main.rs (hunk 4 above)
Comment (Eraden) [reply src/main.rs:375]: `log::debug!("Temperature didn't change");` is followed by `continue`

File: src/main.rs (hunk 3 above)
Comment (Eraden) [reply src/main.rs]: Fixed

**Comment by BoostCookie** [comment]
If @Eraden agrees with my commits I think we should merge this now and talk about the "which temperature to read"-topic in a separate issue.

**Pull request merged**